tsdb/head.go

```diff
@@ -424,6 +439,52 @@ func NewHead(r prometheus.Registerer, l log.Logger, wal *wal.WAL, opts *HeadOpti
 
 func mmappedChunksDir(dir string) string { return filepath.Join(dir, "chunks_head") }
 
+func (h *Head) ApplyConfig(cfg *config.Config) error {
+	if !h.opts.EnableExemplarStorage {
+		return nil
+	}
+
+	// User wants to 'disable' exemplar storage without restarting Prometheus.
+	if cfg.StorageConfig.ExemplarsConfig.MaxExemplars < 0 {
+		h.exemplars = noopExemplarStorage{}
+		// Reset some metrics.
+		h.exemplarMetrics.exemplarsInStorage.Set(0)
+		h.exemplarMetrics.seriesWithExemplarsInStorage.Set(0)
+		h.exemplarMetrics.maxExemplars.Set(-1)
+		return nil
+	}
+
+	// Ugly hack to get the previous default size.
+	if cfg.StorageConfig.ExemplarsConfig.MaxExemplars == 0 {
+		cfg.StorageConfig.ExemplarsConfig.MaxExemplars = config.DefaultExemplarsConfig.MaxExemplars
+	}
```

Comment:
Look at my other comment on the config being a pointer. I think 0 is a valid config to be allowed and can get confusing if 0 turns out to be default. To have the default config users should just not mention anything in the config file.

Reply:
`To have the default config users should just not mention anything in the config file.` This is what I wanted but I really don't understand the config file parsing enough to know how to get the default exemplar storage config passed in only if the CLI parameter is also set. Do you think that's an issue?

cc @roidelapluie

Reply:
Setting the config as a pointer should do the trick of knowing if the config is set in the file or not. See https://github.com/prometheus/prometheus/pull/8974#discussion_r665057394